lib/segment/src/index/field_index/on_disk_map_index.rs

```diff
@@ -0,0 +1,381 @@
+                self.store_cf_name
+            ))
+        })?;
+        Ok(store_ref.flush_cf(cf_handle)?)
+    }
+
+    pub fn match_cardinality(&self, value: &N) -> CardinalityEstimation {
+        let values_count = match self.map.get(value) {
+            None => 0,
+            Some(points) => points.len(),
+        };
+
+        CardinalityEstimation {
+            primary_clauses: vec![],
+            min: values_count,
+            exp: values_count,
+            max: values_count,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn get_values(&self, idx: PointOffsetType) -> Option<&Vec<N>> {
+        self.point_to_values.get(idx as usize)
+    }
+
+    fn add_many_to_map(&mut self, idx: PointOffsetType, values: Vec<N>) -> OperationResult<()> {
+        let store_ref = self.store.borrow();
+        let cf_handle = store_ref.cf_handle(&self.store_cf_name).ok_or_else(|| {
+            OperationError::service_error(&format!(
+                "Index add error: column family {} not found",
+                self.store_cf_name
+            ))
+        })?;
+
+        if self.point_to_values.len() <= idx as usize {
+            self.point_to_values.resize(idx as usize + 1, Vec::new())
+        }
+        self.point_to_values[idx as usize] = values.into_iter().collect();
+        let mut empty = true;
+        for value in &self.point_to_values[idx as usize] {
+            let entry = self.map.entry(value.clone()).or_default();
+            entry.push(idx);
+            empty = false;
+
+            let db_record = Self::encode_db_record(value, idx);
+            store_ref
+                .put_cf_opt(cf_handle, &db_record, &[], &db_write_options())
+                .map_err(|e| {
+                    OperationError::service_error(&format!("Index db update error: {}", e))
+                })?;
+        }
+        if !empty {
+            self.total_points += 1;
+        }
+        Ok(())
+    }
+
+    fn get_iterator(&self, value: &N) -> Box<dyn Iterator<Item = PointOffsetType> + '_> {
+        self.map
+            .get(value)
+            .map(|ids| Box::new(ids.iter().copied()) as Box<dyn Iterator<Item = PointOffsetType>>)
+            .unwrap_or_else(|| Box::new(iter::empty::<PointOffsetType>()))
+    }
+
+    fn encode_db_record(value: &N, idx: PointOffsetType) -> String {
+        format!("{}/{}", value, idx)
+    }
+
+    fn decode_db_record(s: &str) -> OperationResult<(N, PointOffsetType)> {
+        const DECODE_ERR: &str = "Index db parsing error: wrong data format";
+        let separator_pos = s
+            .rfind('/')
+            .ok_or_else(|| OperationError::service_error(DECODE_ERR))?;
+        if separator_pos == s.len() - 1 {
+            return Err(OperationError::service_error(DECODE_ERR));
+        }
+        let value_str = &s[..separator_pos];
+        let value =
+            N::from_str(value_str).map_err(|_| OperationError::service_error(DECODE_ERR))?;
```

Comment:
an alternative approach here could be to use a binary format instead of a string representation, i.e. use u32::to_be_bytes() instead of the display value "{}"

Reply:
let's postpone this change until we have some insights regarding bottlenecks, cause it might be easier to debug with displayed values